PBKDF2 one-shot

vcsjones opened the pull request:
Work left to be done:

- [x] Wait for API approval and feedback (and modify PR accordingly)
    - [x] XML documentation (TBD after approval)
- [x] Handle other platforms accordingly (Android, Browser)
- [x] Comprehensive tests and edge cases
    - [x] Zero length values (empty password bytes, etc)
    - [x] Overlapping buffers
- [x] Perf numbers on throughput and allocations.
- [x] We should use the throwing (non-substituting) version of UTF-8 when performing the chars -> bytes
- [x] Use `BCryptKeyDerivation` for Windows 8+.

Implementation notes:

* In #24897 I (incorrectly) determined that OpenSSL 1.0 did not have `PKCS5_PBKDF2_HMAC`. This was a PEBCAK error while reading documentation. So no shimming required.
* MacOS, Linux, and Windows 7 should not allocate, assuming you are not using one of the allocating helpers. Windows 8+ will mostly be non-allocating. They will allocate a handle, though it can be removed. I am doing some research to determine if removing the handle allocation makes any material impact on real world use cases.

Closes #24897

vcsjones committed:
Basic implementation of one-shot PBKDF2 for all platforms.

vcsjones committed:
Use HashAlgorithmName to PAL

vcsjones committed:
Use psuedo handles if available on the Windows platform.

vcsjones commented:
Unrelated but exciting: All of the macOS work was done on a M1 Mac Mini without Rosetta (running as ARM64). Very cool to see dotnet/runtime in good shape for Apple Silicon.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/ref/System.Security.Cryptography.Algorithms.cs:
```diff
@@ -587,9 +587,14 @@ public partial class Rfc2898DeriveBytes : System.Security.Cryptography.DeriveByt
+        public static byte[] Pbkdf2DeriveBytes(string password, byte[] salt, int iterations, int length, System.Security.Cryptography.HashAlgorithmName hashAlgorithm) { throw null; }
         public override void Reset() { }
     }
-    [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
```
Oops, need to revert this.

vcsjones committed:
Fix compilation

vcsjones committed:
Undo inadvertent change in ref file.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs:
```diff
@@ -0,0 +1,134 @@
+    {
+        private const int OneShotMinIterations = 1000;
+        private const int OneShotMinSaltLength = 128 / 8; // 128-bits
+        private const int OneShotMinExtractLength = 112 / 8; // 114-bits
```
112, or 114?

vcsjones replied:
Ha. 112, which is 14 bytes, and I guess I just kind of smooshed together in the comment.

vcsjones replied:
Also, currently investigating: I think the minimum checks can go away. The way we are hitting OpenSSL, it might not be forcing the check.

vcsjones replied:
Yep, all the limits were able to be lifted. The salt length limitation is retained from the instance based one, though.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs:
```diff
@@ -0,0 +1,134 @@
+                throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmName));
+            }
+
+            if (destination.IsEmpty)
```
This shouldn't be reachable, given the check against OneShotMinExtractLength, right?  An assert might be better if you're just trying to avoid passing NULL to the native layers.

vcsjones replied:
There is no limit on the extract length anymore, so I think this should be retained. The implementations assert `destination` is not empty now.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_evp.h:
```diff
@@ -120,3 +120,12 @@ GetMaxMdSize
 */
 PALEXPORT int32_t CryptoNative_GetMaxMdSize(void);
+
+PALEXPORT int32_t CryptoNative_Pkcs5Pbkdf2Hmac(const char* password,
```
In the original shim we used names like this, because we were just directly shimming the native functions.

In newer operations I've gone for more naming the operation whatever I felt was appropriate.  I'd call this `CryptoNative_Pbkdf2`, for example.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/tests/Rfc2898OneShotTests.cs:
```diff
@@ -0,0 +1,77 @@
+            HashAlgorithmName hashAlgorithmName = new HashAlgorithmName(hashAlgorithm);
+            using Rfc2898DeriveBytes instanceKdf = new Rfc2898DeriveBytes(s_passwordBytes, s_salt, iterations: 1000, hashAlgorithmName);
+            byte[] key1 = instanceKdf.GetBytes(s_extractLength);
+            byte[] key2 = Rfc2898DeriveBytes.Pbkdf2DeriveBytes(
+                s_passwordBytes, s_salt, iterations: 1000, s_extractLength, hashAlgorithmName);
```
I'd move this outside of the scope of instanceKdf, just so it's clear nothing is leaking across.

vcsjones commented:
Hm. Trying to use the pseudo handles in Windows 7 actually crashes the process with an access violation, so we can't even get an NTSTATUS back saying it failed. Do we just do an OS version check?

vcsjones committed:
Detect Windows 7 instead of relying on the NTSTATUS.

vcsjones committed:
Loosen restrictions and add deeper tests

vcsjones committed:
Fix empty password on macOS.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs:
```diff
@@ -0,0 +1,128 @@
+            Span<byte> destination)
+        {
+            byte[] passwordBytes = CryptoPool.Rent(Encoding.UTF8.GetMaxByteCount(password.Length));
+            int passwordBytesWritten = Encoding.UTF8.GetBytes(password, passwordBytes);
```
@bartonjs is there "danger" here that we want to mitigate? e.g. something that results in invalid UTF8 so the bytes that come out are bogus?

bartonjs replied:
Not that I can think of.  The Encoding.UTF8 shared object is in the "map gibberish to the unknown character" mode.  Consistent garbage in produces consistent garbage out.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs:
```diff
@@ -0,0 +1,128 @@
+            HashAlgorithmName hashAlgorithm,
+            Span<byte> destination)
+        {
+            if (salt.Length < MinimumSaltSize)
```
Maybe we don't want a salt limit? IIRC, one of the blockers from DataProtection moving off their own implementation was they permit any size salt. Perhaps we should do the same, here.

vcsjones replied:
I think this is probably the right way to go, though it is a discrepancy from the instance-based implementation. I went ahead and made the change, but, my current thought is: there should be no minimum. But, if we want to get in the business of minimums, eight was probably too-low of a recommendation, anyway.

vcsjones committed:
Permit empty salts

vcsjones committed:
Fix OpenSSL 1.0 with empty values

vcsjones committed:
Rename exported function

vcsjones committed:
Rename Apple exported function

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native/entrypoints.c:
```diff
@@ -214,7 +214,7 @@ static const Entry s_cryptoNative[] =
     DllImportEntry(CryptoNative_PemReadX509FromBioAux)
     DllImportEntry(CryptoNative_PemWriteBioX509Crl)
-    DllImportEntry(CryptoNative_Pkcs5Pbkdf2Hmac)
+    DllImportEntry(CryptoNative_Pbkdf2)
```
/me nudges the new name up a few lines

vcsjones committed:
Cleanup native PALs.

Reorder to keep alphabetic consistency.

Add comment documentation.

Tighten validation between password and salt inputs
and their length. NULL input requires a zero length.

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 24897-pbkdf2-one-shot

vcsjones committed:
Fix issue with not slicing the rented buffer.

While here, use a stack allocation for small password inputs, which
seems likely.

vcsjones commented:
I did some quick performance benchmarks and the results were surprising. On two different machines, BCrypt's `BCryptDeriveKeyPBKDF2` is slower than what we have now with `GetBytes` (though we can avoid allocations):

|       Method | iterations |      Mean |    Error |   StdDev |
|------------- |----------- |----------:|---------:|---------:|
|       Handle |     100000 | 133.18 ms | 1.984 ms | 1.856 ms |
| PseudoHandle |     100000 | 128.62 ms | 1.756 ms | 1.642 ms |
|     GetBytes |     100000 |  76.72 ms | 0.472 ms | 0.394 ms |

With the benchmark code here: https://github.com/vcsjones/Pbkdf2PsuedoHandleBench

I noted that this is not what DataProtection is using. DataProtection appears to be using `BCryptDeriveKey`. I had assumed the PBKDF2 variant was just a helper wrapper around this one. But maybe not.

I'll give that a shot and see if that gets performance closer to what we expect.

vcsjones commented:
Ah okay. I looked in more detail at the DataProtection implementation. The API I am using is the "Windows 7" preferred option while `BCryptDeriveKey` is preferred for Windows 8+. I'm guessing then that the latter is faster, better, stronger, shinier.

vcsjones committed:
Use BCryptKeyDerivation on Windows 8+.

This improves performance on platforms which support it.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs:
```diff
@@ -0,0 +1,265 @@
+                symmetricKeyMaterialLength = 0;
+                clearSpan = default;
+            }
+            else if (password.Length <= hashBlockSizeBytes)
```
We _could_ let `BCryptGenerateSymmetricKey` do this for us, but it will only do it up to 2048 bytes before we need to manually hash the input anyway. I don't know if we gain much by letting `BCryptGenerateSymmetricKey` for values > block_size && < 2048 instead of just doing it ourselves with the one-shots. I'll put some performance numbers together, if the additional complexity is worth the gain, I can bring it in.

vcsjones committed:
Cleanup and some refactoring for Windows implementation

vcsjones committed:
Use throwing UTF8.

If a string input is invalid UTF8, use an encoding that has a throwing
encoder fallback.

vcsjones committed:
Change to follow approved API.

vcsjones committed:
Tests for overlapping input / output buffers.

vcsjones committed:
Add tests around HMAC block size boundaries.

vcsjones committed:
Fix block sizes.

vcsjones committed:
XML documentation.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs:
```diff
@@ -0,0 +1,270 @@
+{
+    internal partial class Pbkdf2Implementation
+    {
+        private static readonly bool s_usePseudoHandles = OperatingSystem.IsWindowsVersionAtLeast(10, 0, 0);
```
If we agree this is an acceptable way to detect platform support for pseudo handles, I will post-this-PR probably pull this up to `Interop.BCrypt` so that it can be shared with the hashing one-shots which do this detection based on "does a particular exported function exist?" in favor of this.

bartonjs replied:
So long as it's computed to a `static readonly` and it's reused, I don't have a strong opinion on how it's done.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs:
```diff
@@ -0,0 +1,270 @@
+                {
+                    generateKeyStatus = Interop.BCrypt.BCryptGenerateSymmetricKey(
+                        (nuint)BCryptAlgPseudoHandle.BCRYPT_PBKDF2_ALG_HANDLE,
+                        out keyHandle,
```
Just jotting down something that I thought of here.

This is the only meaningful allocation that occurs in the Windows 10 implementation. We _could_ avoid the allocation by forgoing the use of a safe handle and just use an `IntPtr` and free it ourselves. This could... reasonably... be done since the handle will never leave this method.

However, I was unable to determine any real world use case for the one shots where this allocation had any meaningful impact.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs:
```diff
@@ -0,0 +1,300 @@
+            Span<byte> passwordBuffer = maxEncodedSize > MaxPasswordStackSize ?
+                (rentedPasswordBuffer = CryptoPool.Rent(maxEncodedSize)) :
+                stackalloc byte[MaxPasswordStackSize];
+            int passwordBytesWritten = s_throwingUtf8Encoding.GetBytes(password, passwordBuffer);
```
This will throw an `EncoderFallbackException` when given an invalid string. Do we want to wrap it in `CryptographicException` or `ArgumentException`?

GrabYourPitchforks replied:
In theory this should be `ArgumentException`, but my care level is minimal. We're not very consistent ourselves. :)

For example: _System.Formats.Asn1_ allows the EncoderFallbackException to bubble up. _System.Text.Json_ and _System.Formats.Cbor_ rewrap with ArgumentException (though _Cbor_ gives the wrong error text when doing so).

vcsjones replied:
Okay, my current inclination is to leave it as-is until someone more strongly suggests I change it.

vcsjones reviewed (commented):
@bartonjs @GrabYourPitchforks I think this is at a point where it can be looked at in more scrutiny and might nearly resemble something complete, barring some self-review questions.

vcsjones commented:
Apology in advance for changing the PR title and breaking email inbox threading.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.Algorithms/tests/Rfc2898OneShotTests.cs:
```diff
@@ -0,0 +1,321 @@
+                yield return new object[] { hashAlgorithm.Name, 257, 257, "0000000000000000", "0000000000000000" };
+            }
+
+            // Test around HMAC SHA1 and SHA256 block boundaries
```
Hooray for edge case testing!

bartonjs replied:
Was there also a >2k-length password test, just to make sure that we don't simplify the code into producing an error?

vcsjones replied:
@bartonjs good idea, added a test with a large salt and large password.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.Algorithms/tests/Rfc2898OneShotTests.cs:
```diff
@@ -0,0 +1,321 @@
+            }
+        }
+
+        private static HashAlgorithmName[] SupportedHashAlgorithms => new []
```
We should probably add the "official" test cases to this this test file as well. See RFC 6070, Sec. 2.

vcsjones replied:
Good idea. One of them has a high iteration count. I'll see how long it takes CI to do it (it only took my box 2 seconds to run it). If CI takes excessively long I will move that one case to its own `OuterLoop` test.

vcsjones replied:
The highest duration I saw was over 20 seconds on Windows 7, which seems iffy for inner-loop speed. I will move it to outerloop.

GrabYourPitchforks reviewed (commented):
Reviewed the Windows implementation and skimmed the non-Windows implementations. Generally LGTM. Some open questions about whether avoiding the allocation is worth it in the Windows case.

vcsjones commented on src/libraries/Common/src/Interop/Windows/BCrypt/Interop.BCryptGenerateSymmetricKey.cs:
```diff
@@ -0,0 +1,32 @@
+        [DllImport(Libraries.BCrypt, CharSet = CharSet.Unicode)]
+        internal static unsafe extern NTSTATUS BCryptGenerateSymmetricKey(
+            nuint hAlgorithm,
+            out SafeBCryptKeyHandle phKey,
```
> all pseudohandle / raw ptr

Hm, my thought was this is okay. The pseudo handles only apply to algorithm handles, not key-instance handles. A pseudo key-handle doesn't make much sense. However a key handle has a lifetime that needs to be cleaned up (unlike the algorithm handle).

bartonjs replied:
I agree with @vcsjones ... the pseudo-handle is input, but phKey is an out parameter.  Sure, it's only going to live long enough to extract the key, but I'm always more comfortable tracking lifetime with SafeHandle than with manual code.

Is it worth using the pseudo-handle for input, then? Sure.  We don't own the lifetime of that, so don't need to track it, so don't need the SafeHandle.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Unix.cs:
```diff
@@ -0,0 +1,32 @@
+
+            if (result != Success)
+            {
+                Debug.Assert(result == 0, $"Unexpected result {result}");
```
> Are internal implementation failures potentially expected; e.g., OOM?

Unclear exactly what you are asking, but I will attempt to elaborate:

1. The actual OpenSSL function `PKCS5_PBKDF2_HMAC` will return 0 on failure, including if it cannot allocate, and 1 on success.
2. We wrap the OpenSSL function with a shim that returns whatever OpenSSL returns. However, we also return `-1` if any of the inputs are valid, like a `NULL` destination or something. That should never happen since we should be validating that the inputs are sane prior to passing to the shim, the shim is just being defensive.
3. So the check here is saying, if it wasn't successful, it better be because the return is "0" (an error) as opposed to "oh no we somehow let bad inputs get to the shim.

All that said, I think then `Debug.Assert` is the right thing to do here, it's a "we have a problematic code path that get bad inputs get to the shim".

This pattern is fairly common throughout the OpenSSL shims.

GrabYourPitchforks replied:
> So the check here is saying, if it wasn't successful, it better be because the return is "0" (an error) as opposed to "oh no we somehow let bad inputs get to the shim".

This clarifies things - thanks!

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs:
```diff
@@ -0,0 +1,300 @@
+            Span<byte> passwordBytes = passwordBuffer.Slice(0, passwordBytesWritten);
+
+            Pbkdf2Core(passwordBytes, salt, destination, iterations, hashAlgorithm);
+            CryptographicOperations.ZeroMemory(passwordBytes);
```
Ohhh good catch.

vcsjones committed:
Code review feedback.

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 24897-pbkdf2-one-shot

vcsjones committed:
Move high iterations test to OuterLoop.

bartonjs commented on src/libraries/Common/src/Interop/Windows/BCrypt/Interop.Blobs.cs:
```diff
@@ -227,10 +227,27 @@ internal struct BCRYPT_ECCFULLKEY_BLOB
+        ///     NCrypt or BCrypt buffer descriptors
         /// </summary>
-        internal enum NCryptBufferDescriptors : int
+        internal enum NCryptBCryptBufferDescriptors : int
```
Consider `CngBufferDescriptors`, since "CNG" and "NCrypt or BCrypt" are... close enough to synonyms.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs:
```diff
@@ -0,0 +1,270 @@
+
+        // For Windows 7 we will use BCryptDeriveKeyPBKDF2. For Windows 8+ we will use BCryptKeyDerivation
+        // since it has better performance.
+        private static readonly bool s_useKeyDerivation = OperatingSystem.IsWindowsVersionAtLeast(8, 0, 0);
```
Does anyone run Win8?  I'm probably fine with just having Win7+ and Win10+ code, and skipping that 8 could be slightly better.  Especially if it simplifies things.  (If it's just this one piece of detection, fine.  But if we have "older API, newer API with SafeHandles, newer API with PseudoHandles" I'd like to eliminate that middle step of complexity)

vcsjones replied:
> and skipping that 8 could be slightly better

@bartonjs the difference between `BCryptKeyDerivation` and `BCryptDeriveKeyPBKDF2` is significant, 1.5x to 2x faster with the former, with common workload parameters. While I would be happy to simplify, it's not a handleful of microseconds. If we wanted to have this so that `DataProtection` could switch, to something in-box, Windows Server 2012 (R2) would take a non-trivial hit.

bartonjs replied:
Sad face.  OK.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs:
```diff
@@ -0,0 +1,270 @@
+
+                    if (openStatus != NTSTATUS.STATUS_SUCCESS)
+                    {
+                        CryptographicOperations.ZeroMemory(clearSpan);
```
```suggestion
                        pbkdf2AlgorithmHandle.Dispose();
                        CryptographicOperations.ZeroMemory(clearSpan);
```

vcsjones replied:
If the resulting NTSTATUS is not `STATUS_SUCCESS`, these "out" safe handles won't be null?

bartonjs replied:
`out SafeHandle` is never null, it'll instead be a SafeHandle tracking IntPtr 0 (or whatever the default ctor does), still registered for finalization (then the finalizer will be like "oh, IsInvalid is true, no need to call ReleaseHandle", but that still meant it was on the finalization queue)

bartonjs replied:
The P/Invoke basically looks like

```C#
// Allocate now so we don't create the native resource then OOM when trying to track it.
safeHandle = new THandle();
CLR_NOTHROW();
IntPtr ptr;
NativeFunction(..., &ptr, ...);
safeHandle.SetHandle(ptr);
CLR_THROWOK();
```

vcsjones replied:
Ah, Спасибо.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs:
```diff
@@ -0,0 +1,270 @@
+
+            if (generateKeyStatus != NTSTATUS.STATUS_SUCCESS)
+            {
+                throw Interop.BCrypt.CreateCryptographicException(generateKeyStatus);
```
```suggestion
                keyHandle.Dispose();
                throw Interop.BCrypt.CreateCryptographicException(generateKeyStatus);
```

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs:
```diff
@@ -0,0 +1,306 @@
+            int iterations,
+            HashAlgorithmName hashAlgorithm)
+        {
+            if (iterations <= 0)
+                throw new ArgumentOutOfRangeException(nameof(iterations), SR.ArgumentOutOfRange_NeedPosNum);
+            if (string.IsNullOrEmpty(hashAlgorithm.Name))
+                throw new ArgumentException(SR.Cryptography_HashAlgorithmNameNullOrEmpty, nameof(hashAlgorithm));
+
```
Ideally this would be done before any of the array-allocating versions allocate their array.  Just so that the OOM for asking for too large of an output on a constrained system doesn't happen before noticing that there were invalid inputs.  Then this method should just assert (typically true of a `*Core` method anyways)

vcsjones replied:
@bartonjs ah, yeah, makes sense. Refactored things a bit to make that work.

vcsjones committed:
Dispose of handles in error conditions

vcsjones committed:
Rename buffer descriptors with clearer prefix

vcsjones committed:
Refactor input validation.

Checks that inputs are valid before doing any allocations.

vcsjones committed:
Test for large passwords and salts.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs:
```diff
@@ -0,0 +1,272 @@
+                        hashBufferSize = SHA512.HashData(password, hashBuffer);
+                        break;
+                    default:
+                        throw new CryptographicException();
```
Hash algorithm selection was already filtered, right?  Let's throw in a Debug.Fail that reminds readers of that, so it's not just a "huh, weird, a default-message throw".

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 24897-pbkdf2-one-shot

vcsjones committed:
Use asserts to make what should be pre-validated paths clearer.
